File: playwright_attendance.js
// playwright_attendance.js
// Node + Playwright: improved DOM-aware extractor for the MGIT portal

const { chromium } = require('playwright');
const twilio = require('twilio');

const LOGIN_URL = 'https://mgit.winnou.net/index.php';
const DASHBOARD_URL = 'https://mgit.winnou.net/index.php';

const COLLEGE_USER = process.env.COLLEGE_USER;
const COLLEGE_PASS = process.env.COLLEGE_PASS;
const MY_PHONE_NUMBER = process.env.MY_PHONE_NUMBER;           // e.g. +91XXXXXXXXXX
const TWILIO_PHONE_NUMBER = process.env.TWILIO_PHONE_NUMBER; // e.g. +1415XXXXXXX
const TWILIO_SID = process.env.TWILIO_SID;
const TWILIO_TOKEN = process.env.TWILIO_TOKEN;

if (!COLLEGE_USER || !COLLEGE_PASS || !MY_PHONE_NUMBER || !TWILIO_PHONE_NUMBER || !TWILIO_SID || !TWILIO_TOKEN) {
  console.error('Missing required environment variables. Exiting.');
  process.exit(1);
}

function numericOk(n) {
  return !isNaN(n) && n >= 0 && n <= 100;
}

(async () => {
  const browser = await chromium.launch({ headless: true, args: ['--no-sandbox'] });
  const context = await browser.newContext();
  const page = await context.newPage();

  try {
    await page.goto(LOGIN_URL, { waitUntil: 'domcontentloaded', timeout: 30000 });

    // Fill username/password by trying several selectors
    const usernameSelectors = [
      'input[type="text"]',
      'input[type="email"]',
      'input[name*=user]',
      'input[name*=email]',
      'input[id*=user]',
      'input[id*=email]',
      'input[name*=username]',
      'input[id*=username]'
    ];
    const passwordSelectors = [
      'input[type="password"]',
      'input[name*=pass]',
      'input[id*=pass]',
      'input[name*=pwd]',
      'input[id*=pwd]'
    ];

    let filledUser = false;
    for (const sel of usernameSelectors) {
      const el = await page.$(sel);
      if (el) {
        await el.fill(COLLEGE_USER);
        filledUser = true;
        break;
      }
    }

    let filledPass = false;
    for (const sel of passwordSelectors) {
      const el = await page.$(sel);
      if (el) {
        await el.fill(COLLEGE_PASS);
        filledPass = true;
        break;
      }
    }

    // Submit: try pressing Enter then try clicking submit button if needed
    if (filledPass) {
      await page.keyboard.press('Enter').catch(()=>{});
    }
    // small wait for possible JS submit
    await page.waitForTimeout(1500);

    // If still on login page, try clicking submit buttons
    if ((await page.url()).includes('index.php')) {
      const submitBtn = await page.$('button[type="submit"], input[type="submit"], a[href*="login"], button:has-text("Login"), button:has-text("Sign in")');
      if (submitBtn) {
        await submitBtn.click().catch(()=>{});
      }
    }

    // Wait for navigation and rendering
    await page.waitForLoadState('networkidle', { timeout: 20000 }).catch(()=>{});
    await page.waitForTimeout(1500);

    // Ensure dashboard loaded - try direct navigate if necessary
    if ((await page.url()).includes('index.php')) {
      await page.goto(DASHBOARD_URL, { waitUntil: 'domcontentloaded' }).catch(()=>{});
      await page.waitForTimeout(1500);
    }

    // --------- DOM-aware extraction ---------
    const result = await page.evaluate(() => {
      // helper inside page
      function isDateContext(text, idx, len) {
        const before = text.slice(Math.max(0, idx - 6), idx);
        const after = text.slice(idx + len, idx + len + 6);
        return before.includes('-') || after.includes('-') || before.includes('/') || after.includes('/');
      }

      function findNumericTokens(text) {
        const re = /(\d{1,3}(?:\.\d{1,2})?)/g;
        const out = [];
        let m;
        while ((m = re.exec(text)) !== null) {
          out.push({ token: m[1], index: m.index, len: m[1].length });
        }
        return out;
      }

      // 1) Try to find elements that contain the word "Attendance"
      const attNodes = Array.from(document.querySelectorAll('body *')).filter(n => {
        try {
          return /\bAttendance\b/i.test(n.innerText || '');
        } catch (e) { return false; }
      });

      const pageText = document.body.innerText || '';

      for (const node of attNodes) {
        // search inside closest meaningful container: node, node.parentElement, node.closest('td,div,section,table')
        const containers = [];
        containers.push(node);
        if (node.parentElement) containers.push(node.parentElement);
        const closest = node.closest('td,div,section,article,table,tbody') || node.parentElement;
        if (closest) containers.push(closest);

        // add siblings of parent to the search space (sometimes value is in sibling cell)
        if (node.parentElement && node.parentElement.parentElement) {
          const siblings = Array.from(node.parentElement.parentElement.children || []);
          containers.push(...siblings);
        }

        // dedupe containers
        const uniqContainers = Array.from(new Set(containers));

        for (const c of uniqContainers) {
          const text = c.innerText || '';
          // a) look for explicit percent first
          const pctMatch = text.match(/(\d{1,3}(?:\.\d{1,2})?)\s*%/);
          if (pctMatch && Number(pctMatch[1]) <= 100) {
            return { found: true, value: (pctMatch[1].replace(/\.?0+$/,'') + '%'), method: 'percent-in-container' };
          }
          // b) look for parentheses like (72.29)
          const parenMatch = text.match(/\((\d{1,3}(?:\.\d{1,2})?)\)/);
          if (parenMatch && Number(parenMatch[1]) <= 100) {
            return { found: true, value: (parenMatch[1].replace(/\.?0+$/,'') + '%'), method: 'paren-in-container' };
          }
          // c) find numeric tokens in container and pick sensible one not part of date
          const tokens = findNumericTokens(text);
          if (tokens.length) {
            for (const t of tokens) {
              if (isDateContext(text, t.index, t.len)) continue;
              const num = Number(t.token);
              if (!isNaN(num) && num >= 0 && num <= 100) {
                return { found: true, value: (String(num).replace(/\.?0+$/,'') + '%'), method: 'token-in-container' };
              }
            }
          }
        }
      }

      // 2) If not found via Attendance label, search for parentheses anywhere but avoid dates
      const allText = document.body.innerText || '';
      const parenAll = Array.from(allText.matchAll(/\((\d{1,3}(?:\.\d{1,2})?)\)/g)).map(m => ({v: m[1], idx: m.index}));
      if (parenAll.length) {
        for (const p of parenAll) {
          if (!isDateContext(allText, p.idx, String(p.v).length)) {
            const n = Number(p.v);
            if (!isNaN(n) && n >= 0 && n <= 100) {
              return { found: true, value: (String(n).replace(/\.?0+$/,'') + '%'), method: 'paren-anywhere' };
            }
          }
        }
      }

      // 3) search for explicit percent anywhere and choose one nearest to the word "Attendance"
      const pctAll = Array.from(allText.matchAll(/(\d{1,3}(?:\.\d{1,2})?)\s*%/g)).map(m => ({v: m[1], idx: m.index}));
      if (pctAll.length) {
        const attIdx = allText.toLowerCase().indexOf('attendance');
        if (attIdx >= 0) {
          pctAll.sort((a,b) => Math.abs(a.idx - attIdx) - Math.abs(b.idx - attIdx));
          const n = Number(pctAll[0].v);
          if (!isNaN(n) && n >= 0 && n <= 100) return { found: true, value: (String(n).replace(/\.?0+$/,'') + '%'), method: 'pct-anywhere-near-att' };
        } else {
          // fallback pick largest sensible pct
          const nums = pctAll.map(x => Number(x.v)).filter(x => !isNaN(x) && x >= 0 && x <= 100);
          if (nums.length) {
            const n = Math.max(...nums);
            return { found: true, value: (String(n).replace(/\.?0+$/,'') + '%'), method: 'pct-anywhere-largest' };
          }
        }
      }

      return { found: false, snippet: allText.slice(0, 2000) };
    });

    if (result && result.found) {
      console.log('EXTRACTED_ATTENDANCE', result.value, 'method=' + (result.method || 'unknown'));
      // send Twilio message
      const client = twilio(TWILIO_SID, TWILIO_TOKEN);
      const body = `📢 College Attendance Update\nYour current attendance is: ${result.value}\n🕒 Have a great day!`;
      const msg = await client.messages.create({
        body,
        from: `whatsapp:${TWILIO_PHONE_NUMBER}`,
        to: `whatsapp:${MY_PHONE_NUMBER}`
      });
      console.log('SENT_SID', msg.sid);
      await browser.close();
      process.exit(0);
    }

    // If extraction failed - log snippet for debugging
    console.error('DEBUG: Unable to extract attendance automatically.');
    if (result && result.snippet) {
      console.error('PAGE_SNIPPET:', result.snippet);
    } else {
      const txt = await page.textContent('body').catch(()=>'');
      console.error('PAGE_SNIPPET:', (txt || '').slice(0, 2000));
    }

    await browser.close();
    process.exit(2);
  } catch (err) {
    console.error('ERROR:', err);
    await browser.close();
    process.exit(1);
  }
})();

File: playwright_attendance.py
# playwright_scripts/playwright_attendance.py
import os
import asyncio
import json
import time
import traceback
from pathlib import Path
from playwright.async_api import async_playwright, TimeoutError as PWTimeout

# output files
DIAGNOSTIC_HTML = Path("diagnostic_page.html")
DIAGNOSTIC_SCREENSHOT = Path("diagnostic_screenshot.png")
FATAL_HTML = Path("fatal_error_page.html")
FATAL_SCREENSHOT = Path("fatal_error_screenshot.png")
RESULT_JSON = Path("result.json")

COL_USER = os.getenv("COLLEGE_USER")
COL_PASS = os.getenv("COLLEGE_PASS")
BASE_URL = os.getenv("COLLEGE_URL", "https://mgit.winnou.net/index.php")
MAX_RETRIES = 3

async def fetch_attendance():
    # basic result template
    result = {"success": False, "attendance": None, "error": None}
    if not COL_USER or not COL_PASS:
        result["error"] = "Missing COLLEGE_USER or COLLEGE_PASS environment variables"
        RESULT_JSON.write_text(json.dumps(result))
        return result

    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=True, args=["--no-sandbox"])
        page = await browser.new_page()
        try:
            # Try a few attempts for flaky network/pages
            for attempt in range(1, MAX_RETRIES + 1):
                try:
                    await page.goto(BASE_URL, timeout=90000)  # 90s
                    await page.wait_for_load_state("networkidle", timeout=45000)
                    break
                except PWTimeout:
                    if attempt == MAX_RETRIES:
                        raise
                    else:
                        await page.screenshot(path=f"retry_nav_{attempt}.png")
                        time.sleep(2 ** attempt)
            # Save initial diagnostic HTML + screenshot
            DIAGNOSTIC_HTML.write_text(await page.content(), encoding="utf-8")
            await page.screenshot(path=str(DIAGNOSTIC_SCREENSHOT), full_page=True)

            # Detect simple CAPTCHA / challenge markers
            # (the login page includes #captcha input; also look for 'captcha' keywords)
            content_lower = (await page.content()).lower()
            if ("captcha" in content_lower) or (await page.query_selector("#captcha") is not None):
                result["error"] = "CAPTCHA detected on login page; cannot automate."
                RESULT_JSON.write_text(json.dumps(result))
                return result

            # Fill login form (selectors observed in diagnostic HTML)
            # update these selectors if site changes
            await page.fill("#username", COL_USER)
            await page.fill("#modlgn_passwd", COL_PASS)

            # Click sign-in button (class 'signinbutton' in uploaded HTML)
            sign_button = await page.query_selector("input.signinbutton") or await page.query_selector("input[type=submit]")
            if not sign_button:
                raise Exception("Sign-in button not found; selector may have changed.")

            # Submit and wait
            await sign_button.click()
            # Wait for navigation / dashboard to load
            try:
                await page.wait_for_load_state("networkidle", timeout=45000)
            except PWTimeout:
                # allow extra time for slow pages
                await page.wait_for_timeout(5000)

            # Save post-login HTML for debugging
            post_html = await page.content()
            Path("after_login_page.html").write_text(post_html, encoding="utf-8")
            await page.screenshot(path="after_login_screenshot.png", full_page=True)

            # Now try to find attendance element — you must replace the selector below
            # with the actual selector shown on the logged-in page.
            # Example strategies:
            # - Search for text "attendance" in the page
            # - Look for a known CSS id/class
            page_text = (await page.inner_text("body")).lower()

            # Heuristic: look for percent numbers near 'attendance'
            if "attendance" in page_text:
                # crude extract: find substring around 'attendance'
                idx = page_text.find("attendance")
                snippet = page_text[max(0, idx-200): idx+200]
                # try to find a number like 85% in snippet
                import re
                m = re.search(r"(\d{1,3}\s*%|\d{1,3}\.\d\s*%)", snippet)
                attendance_val = m.group(0) if m else None
                if not attendance_val:
                    # fallback: return the snippet so you can inspect manually
                    attendance_val = "attendance found in snippet: " + snippet.strip().replace("\n"," ")
                result.update({"success": True, "attendance": attendance_val})
                RESULT_JSON.write_text(json.dumps(result, ensure_ascii=False))
                return result
            else:
                # No attendance text found — maybe we need to navigate to "Attendance" page
                # try clicking the 'Attendance' link that appears in diagnostic tags (if present)
                att_link = await page.query_selector("a:has-text('Attendance')")
                if att_link:
                    await att_link.click()
                    await page.wait_for_load_state("networkidle")
                    await page.screenshot(path="attendance_page_screenshot.png", full_page=True)
                    page_text2 = (await page.inner_text("body")).lower()
                    idx = page_text2.find("attendance")
                    snippet = page_text2[max(0, idx-200): idx+200] if idx!=-1 else page_text2[:400]
                    import re
                    m = re.search(r"(\d{1,3}\s*%|\d{1,3}\.\d\s*%)", snippet)
                    attendance_val = m.group(0) if m else "attendance page loaded but no percent found"
                    result.update({"success": True, "attendance": attendance_val})
                    RESULT_JSON.write_text(json.dumps(result, ensure_ascii=False))
                    return result

            # nothing found
            result["error"] = "Unable to locate attendance text after login. See after_login_page.html and screenshots."
            RESULT_JSON.write_text(json.dumps(result))
            return result

        except Exception as e:
            # fatal: save everything for diagnostics
            err = traceback.format_exc()
            FATAL_HTML.write_text(await page.content() if page else str(err), encoding="utf-8")
            with open("fatal_error.txt", "w", encoding="utf-8") as f:
                f.write(err)
            try:
                await page.screenshot(path=str(FATAL_SCREENSHOT), full_page=True)
            except Exception:
                pass
            result["error"] = f"Exception: {str(e)}; see fatal_error.txt"
            RESULT_JSON.write_text(json.dumps(result))
            return result
        finally:
            await browser.close()

if __name__ == "__main__":
    res = asyncio.run(fetch_attendance())
    print(json.dumps(res, ensure_ascii=False))
    if not res.get("success"):
        # non-zero exit to indicate Playwright failure
        raise SystemExit(2)
    else:
        # success
        raise SystemExit(0)
